File: packages/brain-runtime/src/collaboration.ts
/**
 * LLM-native collaboration helpers for agent-to-agent coordination.
 * These functions use live ACP context to suggest who should speak next,
 * draft a natural message, and compress thread state for the UI/runtime.
 */

import { message as buildMessage } from "@maia/acp";
import type {
  ACPCapabilities,
  ACPEvent,
  ACPMessage,
  AgentPresence,
  MessageIntent,
} from "@maia/acp";
import { callLLMJson } from "./llm";
import type { LLMCallResult } from "./llm";
import type { AgentDefinition, LLMConfig } from "./types";

export interface CollaborationParticipant {
  agentId: string;
  name: string;
  role?: string;
  description?: string;
  skills?: string[];
  acceptsIntents?: MessageIntent[];
  presence?: AgentPresence;
}

export interface CollaborationContext {
  objective: string;
  currentAgentId: string;
  participants: CollaborationParticipant[];
  events: ACPEvent[];
  threadId?: string;
  taskId?: string;
  taskTitle?: string;
  maxWords?: number;
}

export interface ConversationMove {
  action: "reply" | "ask" | "handoff" | "review" | "summarize" | "wait";
  toAgentId?: string;
  intent: MessageIntent;
  reason: string;
  requiresAck?: boolean;
  mentions?: string[];
  confidence?: number;
}

export interface MessageDraft {
  toAgentId: string;
  intent: MessageIntent;
  content: string;
  thinking?: string;
  mood?: ACPMessage["mood"];
  taskId?: string;
  taskTitle?: string;
  threadId?: string;
  requiresAck?: boolean;
  mentions?: string[];
}

export interface ThreadDigest {
  summary: string;
  decisions: string[];
  blockers: string[];
  openQuestions: string[];
  nextAction?: string;
  nextOwnerAgentId?: string;
}

function emptyCost(llm: LLMConfig): LLMCallResult {
  return {
    text: "",
    tokensUsed: 0,
    costUsd: 0,
    model: llm.model ?? "gpt-4o",
    success: true,
  };
}

function normalizeParticipants(
  participants: Array<CollaborationParticipant | AgentDefinition | ACPCapabilities>,
): CollaborationParticipant[] {
  return participants.map((participant) => {
    if ("agent_id" in participant) {
      return {
        agentId: participant.agent_id,
        name: participant.name,
        role: participant.role,
        description: participant.description,
        skills: participant.skills?.map((skill) => skill.description ?? skill.skill_id) ?? [],
        acceptsIntents: participant.accepts_intents,
        presence: participant.presence,
      };
    }
    if ("id" in participant) {
      return {
        agentId: participant.id,
        name: participant.name,
        role: participant.role,
        description: participant.instructions,
        skills: participant.tools ?? [],
      };
    }
    return participant;
  });
}

function truncate(text: string, maxLength: number): string {
  const value = String(text || "").trim();
  if (value.length <= maxLength) {
    return value;
  }
  return `${value.slice(0, Math.max(0, maxLength - 3))}...`;
}

function eventLine(event: ACPEvent): string {
  if (event.event_type === "message") {
    const payload = event.payload as ACPMessage;
    const context = payload.context ?? {};
    const target = payload.to === "agent://broadcast" ? "everyone" : payload.to;
    return [
      `${payload.from} -> ${target}`,
      `[${payload.intent}]`,
      truncate(payload.content, 240),
      context.task_title ? `(task: ${context.task_title})` : "",
      context.delivery_status ? `(delivery: ${context.delivery_status})` : "",
    ]
      .filter(Boolean)
      .join(" ");
  }
  if (event.event_type === "handoff") {
    const payload = event.payload as Record<string, any>;
    return `${payload.from} handed off to ${payload.to}: ${truncate(payload.task?.description ?? "", 220)}`;
  }
  if (event.event_type === "review") {
    const payload = event.payload as Record<string, any>;
    return `${payload.reviewer} reviewed ${payload.author}: ${payload.verdict}`;
  }
  return `${event.agent_id} ${event.event_type}`;
}

function formatParticipants(participants: CollaborationParticipant[], currentAgentId: string): string {
  return participants
    .map((participant) => {
      const intents =
        participant.acceptsIntents && participant.acceptsIntents.length > 0
          ? participant.acceptsIntents.join(", ")
          : "any";
      const presence = participant.presence;
      return [
        `- ${participant.agentId}${participant.agentId === currentAgentId ? " (current speaker)" : ""}`,
        `${participant.name}${participant.role ? `, ${participant.role}` : ""}`,
        participant.description ? `- ${truncate(participant.description, 140)}` : "",
        participant.skills && participant.skills.length > 0
          ? `- skills: ${participant.skills.slice(0, 6).join(", ")}`
          : "",
        `- intents: ${intents}`,
        presence?.availability ? `- availability: ${presence.availability}` : "",
        presence?.current_focus ? `- focus: ${truncate(presence.current_focus, 100)}` : "",
        typeof presence?.active_task_count === "number"
          ? `- active tasks: ${presence.active_task_count}`
          : "",
      ]
        .filter(Boolean)
        .join(" ");
    })
    .join("\n");
}

function formatRecentEvents(events: ACPEvent[]): string {
  if (!events.length) {
    return "(no recent conversation)";
  }
  return events.slice(-12).map(eventLine).join("\n");
}

function lastMessageId(events: ACPEvent[]): string | undefined {
  for (let index = events.length - 1; index >= 0; index--) {
    const event = events[index];
    if (event.event_type !== "message") {
      continue;
    }
    const payload = event.payload as ACPMessage;
    if (payload.context?.message_id) {
      return payload.context.message_id;
    }
  }
  return undefined;
}

function moveSystemPrompt(maxWords: number): string {
  return [
    "You coordinate a team of AI workmates.",
    "Decide the most useful next communication move for the current agent.",
    "Use natural human work patterns: ask for verification when needed, hand off when the work should switch owners, summarize when the thread is getting noisy, and wait when no message adds value.",
    "Prefer the best teammate based on role fit, current focus, and availability.",
    `If a message is needed, keep it under ${maxWords} words.`,
    "Return JSON only with keys: action, to_agent_id, intent, reason, requires_ack, mentions, confidence.",
  ].join(" ");
}

function moveUserPrompt(context: CollaborationContext): string {
  return [
    `Objective: ${context.objective}`,
    context.taskTitle ? `Task title: ${context.taskTitle}` : "",
    context.taskId ? `Task id: ${context.taskId}` : "",
    context.threadId ? `Thread id: ${context.threadId}` : "",
    `Current agent: ${context.currentAgentId}`,
    "",
    "Participants:",
    formatParticipants(context.participants, context.currentAgentId),
    "",
    "Recent events:",
    formatRecentEvents(context.events),
  ]
    .filter(Boolean)
    .join("\n");
}

function draftSystemPrompt(maxWords: number): string {
  return [
    "You write natural teammate-to-teammate messages for an AI team.",
    "Sound like a strong human workmate: direct, specific, and useful.",
    "Do not use canned greetings, filler, or robotic recap.",
    "Ground the message in the supplied thread context and task state.",
    "If the move is a handoff, clearly state what is done and what needs the recipient.",
    "If the move is a question, make it crisp and answerable.",
    `Keep the message under ${maxWords} words unless the context requires a short structured summary.`,
    "Return JSON only with keys: content, intent, thinking, mood, to_agent_id, task_id, task_title, thread_id, requires_ack, mentions.",
  ].join(" ");
}

function draftUserPrompt(context: CollaborationContext, move: ConversationMove): string {
  return [
    `Objective: ${context.objective}`,
    `Current agent: ${context.currentAgentId}`,
    `Planned move: ${move.action}`,
    `Reason: ${move.reason}`,
    move.toAgentId ? `Recipient: ${move.toAgentId}` : "",
    `Intent: ${move.intent}`,
    context.taskTitle ? `Task title: ${context.taskTitle}` : "",
    context.taskId ? `Task id: ${context.taskId}` : "",
    context.threadId ? `Thread id: ${context.threadId}` : "",
    "",
    "Participants:",
    formatParticipants(context.participants, context.currentAgentId),
    "",
    "Recent events:",
    formatRecentEvents(context.events),
  ]
    .filter(Boolean)
    .join("\n");
}

function digestSystemPrompt(): string {
  return [
    "You compress agent coordination threads into an operational digest.",
    "Capture what has been decided, what is blocked, what remains unclear, and who should act next.",
    "Return JSON only with keys: summary, decisions, blockers, open_questions, next_action, next_owner_agent_id.",
  ].join(" ");
}

function digestUserPrompt(context: CollaborationContext): string {
  return [
    `Objective: ${context.objective}`,
    context.taskTitle ? `Task title: ${context.taskTitle}` : "",
    context.threadId ? `Thread id: ${context.threadId}` : "",
    "",
    "Participants:",
    formatParticipants(context.participants, context.currentAgentId),
    "",
    "Recent events:",
    formatRecentEvents(context.events),
  ]
    .filter(Boolean)
    .join("\n");
}

export async function suggestConversationMove(
  llm: LLMConfig,
  rawContext: CollaborationContext,
): Promise<{ move: ConversationMove; cost: LLMCallResult }> {
  const context = {
    ...rawContext,
    participants: normalizeParticipants(rawContext.participants),
  };
  if (context.participants.length <= 1) {
    return {
      move: {
        action: "wait",
        intent: "summarize",
        reason: "No teammate is available for coordination.",
        confidence: 1,
      },
      cost: emptyCost(llm),
    };
  }

  const { data, cost } = await callLLMJson<{
    action?: ConversationMove["action"];
    to_agent_id?: string;
    intent?: MessageIntent;
    reason?: string;
    requires_ack?: boolean;
    mentions?: string[];
    confidence?: number;
  }>(
    llm,
    moveSystemPrompt(rawContext.maxWords ?? 80),
    moveUserPrompt(context),
    {
      action: "wait",
      intent: "summarize",
      reason: "No additional coordination is needed.",
      confidence: 0.5,
    },
  );

  return {
    move: {
      action: data.action ?? "wait",
      toAgentId: data.to_agent_id,
      intent: data.intent ?? "clarify",
      reason: data.reason ?? "No additional coordination is needed.",
      requiresAck: data.requires_ack ?? false,
      mentions: data.mentions ?? [],
      confidence: data.confidence,
    },
    cost,
  };
}

export async function draftConversationMessage(
  llm: LLMConfig,
  rawContext: CollaborationContext,
  move: ConversationMove,
): Promise<{ draft: MessageDraft | null; message: ACPMessage | null; cost: LLMCallResult }> {
  if (move.action === "wait" || !move.toAgentId) {
    return { draft: null, message: null, cost: emptyCost(llm) };
  }

  const context = {
    ...rawContext,
    participants: normalizeParticipants(rawContext.participants),
  };

  const { data, cost } = await callLLMJson<{
    content?: string;
    intent?: MessageIntent;
    thinking?: string;
    mood?: ACPMessage["mood"];
    to_agent_id?: string;
    task_id?: string;
    task_title?: string;
    thread_id?: string;
    requires_ack?: boolean;
    mentions?: string[];
  }>(
    llm,
    draftSystemPrompt(rawContext.maxWords ?? 80),
    draftUserPrompt(context, move),
    {
      content: "",
      intent: move.intent,
      to_agent_id: move.toAgentId,
      task_id: rawContext.taskId,
      task_title: rawContext.taskTitle,
      thread_id: rawContext.threadId,
      requires_ack: move.requiresAck,
      mentions: move.mentions ?? [],
    },
  );

  const content = String(data.content ?? "").trim();
  if (!content) {
    return { draft: null, message: null, cost };
  }

  const draft: MessageDraft = {
    toAgentId: data.to_agent_id ?? move.toAgentId,
    intent: data.intent ?? move.intent,
    content,
    thinking: data.thinking,
    mood: data.mood,
    taskId: data.task_id ?? rawContext.taskId,
    taskTitle: data.task_title ?? rawContext.taskTitle,
    threadId: data.thread_id ?? rawContext.threadId,
    requiresAck: data.requires_ack ?? move.requiresAck,
    mentions: data.mentions ?? move.mentions ?? [],
  };

  return {
    draft,
    message: buildMessage({
      from: rawContext.currentAgentId,
      to: draft.toAgentId,
      intent: draft.intent,
      content: draft.content,
      thinking: draft.thinking,
      mood: draft.mood,
      inReplyTo: lastMessageId(rawContext.events),
      taskId: draft.taskId,
      taskTitle: draft.taskTitle,
      threadId: draft.threadId,
      requiresAck: draft.requiresAck,
      mentions: draft.mentions,
    }),
    cost,
  };
}

export async function summarizeConversationThread(
  llm: LLMConfig,
  rawContext: CollaborationContext,
): Promise<{ digest: ThreadDigest; cost: LLMCallResult }> {
  const context = {
    ...rawContext,
    participants: normalizeParticipants(rawContext.participants),
  };

  const { data, cost } = await callLLMJson<{
    summary?: string;
    decisions?: string[];
    blockers?: string[];
    open_questions?: string[];
    next_action?: string;
    next_owner_agent_id?: string;
  }>(
    llm,
    digestSystemPrompt(),
    digestUserPrompt(context),
    {
      summary: "",
      decisions: [],
      blockers: [],
      open_questions: [],
    },
  );

  return {
    digest: {
      summary: data.summary ?? "",
      decisions: data.decisions ?? [],
      blockers: data.blockers ?? [],
      openQuestions: data.open_questions ?? [],
      nextAction: data.next_action,
      nextOwnerAgentId: data.next_owner_agent_id,
    },
    cost,
  };
}

File: packages/brain-py/maia_brain/collaboration.py
"""LLM-native collaboration helpers for agent-to-agent coordination."""

from __future__ import annotations

import json
import re
from typing import Any, Literal

from pydantic import BaseModel, Field

from maia_acp.builders import message as build_message
from maia_acp.types import ACPEvent, ACPCapabilities, ACPMessage, AgentPresence, MessageIntent
from maia_brain.llm import call_llm
from maia_brain.types import AgentDefinition, LLMConfig, LLMResult


ConversationAction = Literal["reply", "ask", "handoff", "review", "summarize", "wait"]


class CollaborationParticipant(BaseModel):
    agent_id: str = Field(alias="agentId")
    name: str
    role: str | None = None
    description: str | None = None
    skills: list[str] = Field(default_factory=list)
    accepts_intents: list[MessageIntent] = Field(default_factory=list, alias="acceptsIntents")
    presence: AgentPresence | None = None

    model_config = {"populate_by_name": True}


class CollaborationContext(BaseModel):
    objective: str
    current_agent_id: str = Field(alias="currentAgentId")
    participants: list[CollaborationParticipant]
    events: list[ACPEvent] = Field(default_factory=list)
    thread_id: str | None = Field(default=None, alias="threadId")
    task_id: str | None = Field(default=None, alias="taskId")
    task_title: str | None = Field(default=None, alias="taskTitle")
    max_words: int = Field(default=80, alias="maxWords")

    model_config = {"populate_by_name": True}


class ConversationMove(BaseModel):
    action: ConversationAction
    intent: MessageIntent
    reason: str
    to_agent_id: str | None = Field(default=None, alias="toAgentId")
    requires_ack: bool = Field(default=False, alias="requiresAck")
    mentions: list[str] = Field(default_factory=list)
    confidence: float | None = None

    model_config = {"populate_by_name": True}


class MessageDraft(BaseModel):
    to_agent_id: str = Field(alias="toAgentId")
    intent: MessageIntent
    content: str
    thinking: str | None = None
    mood: ACPMessage.model_fields["mood"].annotation | None = None
    task_id: str | None = Field(default=None, alias="taskId")
    task_title: str | None = Field(default=None, alias="taskTitle")
    thread_id: str | None = Field(default=None, alias="threadId")
    requires_ack: bool = Field(default=False, alias="requiresAck")
    mentions: list[str] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class ThreadDigest(BaseModel):
    summary: str
    decisions: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(default_factory=list)
    open_questions: list[str] = Field(default_factory=list, alias="openQuestions")
    next_action: str | None = Field(default=None, alias="nextAction")
    next_owner_agent_id: str | None = Field(default=None, alias="nextOwnerAgentId")

    model_config = {"populate_by_name": True}


class MoveSuggestionResult(BaseModel):
    move: ConversationMove
    cost: LLMResult


class MessageDraftResult(BaseModel):
    draft: MessageDraft | None = None
    message: ACPMessage | None = None
    cost: LLMResult


class ThreadDigestResult(BaseModel):
    digest: ThreadDigest
    cost: LLMResult


def _empty_cost(llm: LLMConfig) -> LLMResult:
    return LLMResult(content="", tokens=0, cost=0.0)


def _truncate(text: str, max_length: int) -> str:
    value = str(text or "").strip()
    if len(value) <= max_length:
        return value
    return f"{value[: max(0, max_length - 3)]}..."


def _normalize_participant(
    participant: CollaborationParticipant | AgentDefinition | ACPCapabilities | dict[str, Any],
) -> CollaborationParticipant:
    if isinstance(participant, CollaborationParticipant):
        return participant
    if isinstance(participant, ACPCapabilities):
        return CollaborationParticipant(
            agent_id=participant.agent_id,
            name=participant.name,
            role=participant.role,
            description=participant.description,
            skills=[skill.description or skill.skill_id for skill in participant.skills],
            accepts_intents=participant.accepts_intents,
            presence=participant.presence,
        )
    if isinstance(participant, AgentDefinition):
        return CollaborationParticipant(
            agent_id=participant.id,
            name=participant.name,
            role=participant.role,
            description=participant.instructions,
            skills=participant.tools,
        )
    return CollaborationParticipant.model_validate(participant)


def _normalize_context(raw_context: CollaborationContext | dict[str, Any]) -> CollaborationContext:
    if isinstance(raw_context, CollaborationContext):
        return CollaborationContext(
            objective=raw_context.objective,
            current_agent_id=raw_context.current_agent_id,
            participants=[_normalize_participant(p) for p in raw_context.participants],
            events=raw_context.events,
            thread_id=raw_context.thread_id,
            task_id=raw_context.task_id,
            task_title=raw_context.task_title,
            max_words=raw_context.max_words,
        )

    payload = dict(raw_context)
    payload["participants"] = [_normalize_participant(p) for p in payload.get("participants", [])]
    return CollaborationContext.model_validate(payload)


def _normalize_move(raw_move: ConversationMove | dict[str, Any]) -> ConversationMove:
    if isinstance(raw_move, ConversationMove):
        return raw_move
    return ConversationMove.model_validate(raw_move)


def _event_line(event: ACPEvent) -> str:
    if event.event_type == "message":
        payload = event.as_message()
        context = payload.context or {}
        target = "everyone" if payload.to == "agent://broadcast" else payload.to
        parts = [
            f"{payload.from_agent} -> {target}",
            f"[{payload.intent}]",
            _truncate(payload.content, 240),
        ]
        if context.get("task_title"):
            parts.append(f"(task: {context['task_title']})")
        if context.get("delivery_status"):
            parts.append(f"(delivery: {context['delivery_status']})")
        return " ".join(part for part in parts if part)

    if event.event_type == "handoff":
        payload = event.as_handoff()
        return f"{payload.from_agent} handed off to {payload.to}: {_truncate(payload.task.description, 220)}"

    if event.event_type == "review":
        payload = event.as_review()
        return f"{payload.reviewer} reviewed {payload.author}: {payload.verdict}"

    return f"{event.agent_id} {event.event_type}"


def _format_participants(participants: list[CollaborationParticipant], current_agent_id: str) -> str:
    lines: list[str] = []
    for participant in participants:
        line_parts = [
            f"- {participant.agent_id}{' (current speaker)' if participant.agent_id == current_agent_id else ''}",
            f"{participant.name}{f', {participant.role}' if participant.role else ''}",
        ]
        if participant.description:
            line_parts.append(f"- {_truncate(participant.description, 140)}")
        if participant.skills:
            line_parts.append(f"- skills: {', '.join(participant.skills[:6])}")
        intents = ", ".join(participant.accepts_intents) if participant.accepts_intents else "any"
        line_parts.append(f"- intents: {intents}")
        presence = participant.presence
        if presence and presence.availability:
            line_parts.append(f"- availability: {presence.availability}")
        if presence and presence.current_focus:
            line_parts.append(f"- focus: {_truncate(presence.current_focus, 100)}")
        if presence and presence.active_task_count is not None:
            line_parts.append(f"- active tasks: {presence.active_task_count}")
        lines.append(" ".join(line_parts))
    return "\n".join(lines)


def _format_recent_events(events: list[ACPEvent]) -> str:
    if not events:
        return "(no recent conversation)"
    return "\n".join(_event_line(event) for event in events[-12:])


def _last_message_id(events: list[ACPEvent]) -> str | None:
    for event in reversed(events):
        if event.event_type != "message":
            continue
        payload = event.as_message()
        message_id = payload.context.get("message_id") if payload.context else None
        if message_id:
            return str(message_id)
    return None


async def _call_llm_json_with_result(
    llm: LLMConfig,
    system_prompt: str,
    user_prompt: str,
    fallback: dict[str, Any],
) -> tuple[dict[str, Any], LLMResult]:
    result = await call_llm(llm, system_prompt, user_prompt)
    try:
        match = re.search(r"[\[{][\s\S]*[\]}]", result.content)
        if match:
            parsed = json.loads(match.group())
            if isinstance(parsed, dict):
                return parsed, result
    except (json.JSONDecodeError, TypeError):
        pass
    return fallback, result


def _move_system_prompt(max_words: int) -> str:
    return " ".join(
        [
            "You coordinate a team of AI workmates.",
            "Decide the most useful next communication move for the current agent.",
            "Use natural human work patterns: ask for verification when needed, hand off when the work should switch owners, summarize when the thread is getting noisy, and wait when no message adds value.",
            "Prefer the best teammate based on role fit, current focus, and availability.",
            f"If a message is needed, keep it under {max_words} words.",
            "Return JSON only with keys: action, to_agent_id, intent, reason, requires_ack, mentions, confidence.",
        ]
    )


def _move_user_prompt(context: CollaborationContext) -> str:
    lines = [
        f"Objective: {context.objective}",
        f"Task title: {context.task_title}" if context.task_title else "",
        f"Task id: {context.task_id}" if context.task_id else "",
        f"Thread id: {context.thread_id}" if context.thread_id else "",
        f"Current agent: {context.current_agent_id}",
        "",
        "Participants:",
        _format_participants(context.participants, context.current_agent_id),
        "",
        "Recent events:",
        _format_recent_events(context.events),
    ]
    return "\n".join(line for line in lines if line)


def _draft_system_prompt(max_words: int) -> str:
    return " ".join(
        [
            "You write natural teammate-to-teammate messages for an AI team.",
            "Sound like a strong human workmate: direct, specific, and useful.",
            "Do not use canned greetings, filler, or robotic recap.",
            "Ground the message in the supplied thread context and task state.",
            "If the move is a handoff, clearly state what is done and what needs the recipient.",
            "If the move is a question, make it crisp and answerable.",
            f"Keep the message under {max_words} words unless the context requires a short structured summary.",
            "Return JSON only with keys: content, intent, thinking, mood, to_agent_id, task_id, task_title, thread_id, requires_ack, mentions.",
        ]
    )


def _draft_user_prompt(context: CollaborationContext, move: ConversationMove) -> str:
    lines = [
        f"Objective: {context.objective}",
        f"Current agent: {context.current_agent_id}",
        f"Planned move: {move.action}",
        f"Reason: {move.reason}",
        f"Recipient: {move.to_agent_id}" if move.to_agent_id else "",
        f"Intent: {move.intent}",
        f"Task title: {context.task_title}" if context.task_title else "",
        f"Task id: {context.task_id}" if context.task_id else "",
        f"Thread id: {context.thread_id}" if context.thread_id else "",
        "",
        "Participants:",
        _format_participants(context.participants, context.current_agent_id),
        "",
        "Recent events:",
        _format_recent_events(context.events),
    ]
    return "\n".join(line for line in lines if line)


def _digest_system_prompt() -> str:
    return " ".join(
        [
            "You compress agent coordination threads into an operational digest.",
            "Capture what has been decided, what is blocked, what remains unclear, and who should act next.",
            "Return JSON only with keys: summary, decisions, blockers, open_questions, next_action, next_owner_agent_id.",
        ]
    )


def _digest_user_prompt(context: CollaborationContext) -> str:
    lines = [
        f"Objective: {context.objective}",
        f"Task title: {context.task_title}" if context.task_title else "",
        f"Thread id: {context.thread_id}" if context.thread_id else "",
        "",
        "Participants:",
        _format_participants(context.participants, context.current_agent_id),
        "",
        "Recent events:",
        _format_recent_events(context.events),
    ]
    return "\n".join(line for line in lines if line)


async def suggest_conversation_move(
    llm: LLMConfig,
    raw_context: CollaborationContext | dict[str, Any],
) -> MoveSuggestionResult:
    context = _normalize_context(raw_context)
    if len(context.participants) <= 1:
        return MoveSuggestionResult(
            move=ConversationMove(
                action="wait",
                intent="summarize",
                reason="No teammate is available for coordination.",
                confidence=1.0,
            ),
            cost=_empty_cost(llm),
        )

    data, cost = await _call_llm_json_with_result(
        llm,
        _move_system_prompt(context.max_words),
        _move_user_prompt(context),
        {
            "action": "wait",
            "intent": "summarize",
            "reason": "No additional coordination is needed.",
            "confidence": 0.5,
        },
    )
    return MoveSuggestionResult(
        move=ConversationMove.model_validate(
            {
                "action": data.get("action", "wait"),
                "to_agent_id": data.get("to_agent_id"),
                "intent": data.get("intent", "clarify"),
                "reason": data.get("reason", "No additional coordination is needed."),
                "requires_ack": data.get("requires_ack", False),
                "mentions": data.get("mentions", []),
                "confidence": data.get("confidence"),
            }
        ),
        cost=cost,
    )


async def draft_conversation_message(
    llm: LLMConfig,
    raw_context: CollaborationContext | dict[str, Any],
    raw_move: ConversationMove | dict[str, Any],
) -> MessageDraftResult:
    context = _normalize_context(raw_context)
    move = _normalize_move(raw_move)
    if move.action == "wait" or not move.to_agent_id:
        return MessageDraftResult(draft=None, message=None, cost=_empty_cost(llm))

    data, cost = await _call_llm_json_with_result(
        llm,
        _draft_system_prompt(context.max_words),
        _draft_user_prompt(context, move),
        {
            "content": "",
            "intent": move.intent,
            "to_agent_id": move.to_agent_id,
            "task_id": context.task_id,
            "task_title": context.task_title,
            "thread_id": context.thread_id,
            "requires_ack": move.requires_ack,
            "mentions": move.mentions,
        },
    )

    content = str(data.get("content", "")).strip()
    if not content:
        return MessageDraftResult(draft=None, message=None, cost=cost)

    draft = MessageDraft.model_validate(
        {
            "to_agent_id": data.get("to_agent_id", move.to_agent_id),
            "intent": data.get("intent", move.intent),
            "content": content,
            "thinking": data.get("thinking"),
            "mood": data.get("mood"),
            "task_id": data.get("task_id", context.task_id),
            "task_title": data.get("task_title", context.task_title),
            "thread_id": data.get("thread_id", context.thread_id),
            "requires_ack": data.get("requires_ack", move.requires_ack),
            "mentions": data.get("mentions", move.mentions),
        }
    )
    message_payload = build_message(
        from_agent=context.current_agent_id,
        to=draft.to_agent_id,
        intent=draft.intent,
        content=draft.content,
        thinking=draft.thinking,
        mood=draft.mood,
        in_reply_to=_last_message_id(context.events),
        task_id=draft.task_id,
        task_title=draft.task_title,
        thread_id=draft.thread_id,
        requires_ack=draft.requires_ack,
        mentions=draft.mentions,
    )
    return MessageDraftResult(
        draft=draft,
        message=ACPMessage.model_validate(message_payload),
        cost=cost,
    )


async def summarize_conversation_thread(
    llm: LLMConfig,
    raw_context: CollaborationContext | dict[str, Any],
) -> ThreadDigestResult:
    context = _normalize_context(raw_context)
    data, cost = await _call_llm_json_with_result(
        llm,
        _digest_system_prompt(),
        _digest_user_prompt(context),
        {
            "summary": "",
            "decisions": [],
            "blockers": [],
            "open_questions": [],
        },
    )
    return ThreadDigestResult(
        digest=ThreadDigest.model_validate(
            {
                "summary": data.get("summary", ""),
                "decisions": data.get("decisions", []),
                "blockers": data.get("blockers", []),
                "open_questions": data.get("open_questions", []),
                "next_action": data.get("next_action"),
                "next_owner_agent_id": data.get("next_owner_agent_id"),
            }
        ),
        cost=cost,
    )

